src/sql/parts/notebook/models/notebookContexts.ts

```diff
@@ -0,0 +1,154 @@
+import { IDefaultConnection, notebookConstants, INotebookModelOptions } from 'sql/parts/notebook/models/modelInterfaces';
+import { INotificationService } from 'vs/platform/notification/common/notification';
+import { IConnectionManagementService } from 'sql/parts/connection/common/connectionManagement';
+import { ConnectionProfile } from 'sql/parts/connection/common/connectionProfile';
+
+export class NotebookContexts {
+	private static MSSQL_PROVIDER = 'MSSQL';
+
+	private static get DefaultContext(): IDefaultConnection {
+		let defaultConnection: ConnectionProfile = <any>{
+			providerName: NotebookContexts.MSSQL_PROVIDER,
+			id: '-1',
+			serverName: localize('selectConnection', 'Select connection')
+		};
+
+		return {
+			// default context if no other contexts are applicable
+			defaultConnection: defaultConnection,
+			otherConnections: [defaultConnection]
+		};
+	}
+
+	private static get LocalContext(): IDefaultConnection {
+		let localConnection: ConnectionProfile = <any>{
+			providerName: NotebookContexts.MSSQL_PROVIDER,
+			id: '-1',
+			serverName: localize('localhost', 'Localhost')
+		};
+
+		return {
+			// default context if no other contexts are applicable
+			defaultConnection: localConnection,
+			otherConnections: [localConnection]
+		};
+	}
+
+	/**
+	 * Get all of the applicable contexts for a given kernel
+	 * @param apiWrapper ApiWrapper
+	 * @param kernelChangedArgs kernel changed args (both old and new kernel info)
+	 * @param profile current connection profile
+	 */
+	public static async getContextsForKernel(connectionService: IConnectionManagementService, connProviderIds: string[], kernelChangedArgs?: nb.IKernelChangedArgs, profile?: IConnectionProfile): Promise<IDefaultConnection> {
+		let connections: IDefaultConnection = this.DefaultContext;
+		if (!profile) {
+			if (!kernelChangedArgs || !kernelChangedArgs.newValue ||
+				(kernelChangedArgs.oldValue && kernelChangedArgs.newValue.id === kernelChangedArgs.oldValue.id)) {
+				// nothing to do, kernels are the same or new kernel is undefined
+				return connections;
+			}
+		}
+		if (kernelChangedArgs && kernelChangedArgs.newValue && kernelChangedArgs.newValue.name && connProviderIds.length < 1) {
+			return connections;
+		} else {
+			connections = await this.getActiveContexts(connectionService, connProviderIds, profile);
+		}
+		return connections;
+	}
+
+	/**
+	 * Get all active contexts and sort them
+	 * @param apiWrapper ApiWrapper
+	 * @param profile current connection profile
+	 */
+	public static async getActiveContexts(connectionService: IConnectionManagementService, connProviderIds: string[], profile: IConnectionProfile): Promise<IDefaultConnection> {
+		let defaultConnection: ConnectionProfile = NotebookContexts.DefaultContext.defaultConnection;
+		let activeConnections: ConnectionProfile[] = await connectionService.getActiveConnections();
+		// If no connections exist, only show 'n/a'
+		if (activeConnections && activeConnections.length > 0) {
```

Comment:
Maybe doc why we have this set of decisions. E.g. pseudocode on what's meant to happen.